presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexFactory.java

```diff
@@ -58,50 +87,64 @@ public PagesSpatialIndexFactory(List<Type> types, List<Type> outputTypes)
         return outputTypes;
     }
 
-    public synchronized ListenableFuture<PagesSpatialIndex> createPagesSpatialIndex()
+    /**
+     * Called by {@link SpatialJoinOperator} to get a Future that provides spatial
+     * index {@link Supplier} when done. {@link SpatialJoinOperator} must call
+     * {@link #probeOperatorFinished()} on completion to signal that spatial index
+     * is no longer needed.
+     */
+    public synchronized ListenableFuture<Supplier<PagesSpatialIndex>> createPagesSpatialIndex()
     {
-        activeProbeOperators++;
+        activeProbeOperators.retain();
 
         if (pagesSpatialIndex != null) {
             return immediateFuture(pagesSpatialIndex);
         }
 
-        SettableFuture<PagesSpatialIndex> future = SettableFuture.create();
+        SettableFuture<Supplier<PagesSpatialIndex>> future = SettableFuture.create();
         pagesSpatialIndexFutures.add(future);
         return future;
     }
 
+    /**
+     * Called by SpatialJoinOperatorFactory to indicate that all
+     * {@link SpatialJoinOperator} have been created.
+     */
     public synchronized void noMoreProbeOperators()
     {
-        activeProbeOperators--;
-        if (activeProbeOperators == 0) {
-            indexNoLongerNeeded.set(null);
-        }
+        activeProbeOperators.release();
     }
 
+    /**
+     * Called by {@link SpatialJoinOperator} on completion to signal that spatial index
+     * is no longer needed.
+     */
     public synchronized void probeOperatorFinished()
     {
-        activeProbeOperators--;
-        if (activeProbeOperators == 0) {
-            indexNoLongerNeeded.set(null);
-        }
+        activeProbeOperators.release();
     }
 
-    public ListenableFuture<?> lendPagesSpatialIndex(PagesSpatialIndex pagesSpatialIndex)
+    /**
+     * Called by {@link SpatialIndexBuilderOperator} to provide spatial index
+     * {@link Supplier} for {@link SpatialJoinOperator}S to use.
+     *
+     * Returns a Future that completes once all the {@link SpatialJoinOperator}S have completed.
+     */
+    public ListenableFuture<?> lendPagesSpatialIndex(Supplier<PagesSpatialIndex> pagesSpatialIndex)
     {
         requireNonNull(pagesSpatialIndex, "pagesSpatialIndex is null");
 
-        List<SettableFuture<PagesSpatialIndex>> settableFutures;
+        List<SettableFuture<Supplier<PagesSpatialIndex>>> settableFutures;
         synchronized (this) {
             this.pagesSpatialIndex = pagesSpatialIndex;
             settableFutures = ImmutableList.copyOf(pagesSpatialIndexFutures);
             pagesSpatialIndexFutures.clear();
         }
 
-        for (SettableFuture<PagesSpatialIndex> settableFuture : settableFutures) {
+        for (SettableFuture<Supplier<PagesSpatialIndex>> settableFuture : settableFutures) {
             settableFuture.set(pagesSpatialIndex);
         }
 
-        return indexNoLongerNeeded;
+        return activeProbeOperators.getFreeFuture();
```

Comment:
When this future is satisfied, you also need to null out `pagesSpatialIndex` to let the memory go.
This also means that if the future is satisfied upon entering this method, you should not store the supplier on the field at all:
```
if (activeProbeOperators.getFreeFuture().isDone()) {
   return NOT_BLOCKED;
}

synchronized {
  verify(this.pagesSpatialIndex == null);
  this.pagesSpatialIndex = pagesSpatialIndex
 ....
```